### ISSUE TITLE
Training, validation, and testing of models, and Predicting using a model

### ISSUE THREAD

**Issue by selcouthlyBlue** [rwth-i6/returnn#7]
I have a few questions in training, validating, and testing of models in the mdlstm demo IAM.

Tl;dr, what are the steps in training, validating, and testing of models? How do I use a model to predict?

1. In training (I used config_real for this btw), where do I find the training error? The same question goes for validating and testing.
2. I have already have my models generated from training. I have also tried validating one of them by using config_fwd and got a validated model from that. Now how do I test this model?
3. Say I got a model from testing. How do I use said model?

**Comment by pvoigtlaender** [comment]
1) during training, after each epoch, in the log file you will find a line similar to this:
epoch 1 score: 2.65181904453 elapsed: 1:33:08 dev: score 1.39069169463 error 0.369806780381
the number after error shows the label error rate on the validation ("dev") set. The train error is currently not printed I think

2)
the config_fwd is not about "validating the model", but it will use the model to produce outputs for the validation data and write it to a file (in this case to "output_file": "mdlstm_real_valid.h5" ).
You can then view the posteriors by using
h5dump mdlstm_real_valid.h5

3)
Using config_fwd, you can produce posteriors for the specified input file (here "eval": "features/raw/valid.h5") as a hdf file. This hdf file can then be decoded e.g. by RASR, Kaldi or a simple self written decoder

**Comment by selcouthlyBlue** [comment]
I did find a line similar to that in log file.

When do I get to use the test set (test.h5)? :O

**Comment by pvoigtlaender** [comment]
the usual procedure is to forward on the testset (using config_fwd and changing the "eval" entry to "eval": "features/raw/valid.h5" and the "output_file" entry to, say  "output_file": "mdlstm_real_test.h5")
Then you can use the posteriors from mdlstm_real_test.h5 for decoding

**Comment by selcouthlyBlue** [comment]
I did some additional reading on RASR and it all makes sense now :O 
Thank you so much for answering my queries 👍

**Comment by selcouthlyBlue** [comment]
Do the scores mean "accuracy" of the model? :O

**Comment by pvoigtlaender** [comment]
Score in this case is the CTC loss function

**Comment by selcouthlyBlue** [comment]
I'm planning to use Kaldi to decode the hdf file as I plan to integrate it to a mobile application. Have you tried using Kaldi for decoding the hdf file?

**Comment by selcouthlyBlue** [comment]
How do I use the model to recognize the handwriting of a single image that contains the handwritten text? :O

**Comment by pvoigtlaender** [comment]
first you create a hdf 5 file containing the data for the image. For this, you can adapt the script https://github.com/rwth-i6/returnn/blob/master/demos/mdlstm/IAM/create_IAM_dataset.py or you might also start from here https://github.com/rwth-i6/returnn/blob/master/demos/mdlstm/artificial/create_test_h5.py and replace the artificial data with your image.

Afterwards you forward your trained model on this data, which gives you a hdf5 file with posteriors for the image.

Then you run a decoder (e.g. Kaldi, RASR).

It might also be possible in a easier way to run the best path decoder using "task": "daemon". For a rough explanation, please view this issue: https://github.com/rwth-i6/returnn/issues/3

**Comment by selcouthlyBlue** [comment]
Thanks for the reply. So far here are the steps I've done.

1. I've created an hdf 5 file containing the data for the image using this code I've adapted from create_IAM_dataset.py.

`def hdf5_strings(handle, name, data):
    try:
        S = max([len(d) for d in data])
        dset = handle.create_dataset(name, (len(data),), dtype="S" + str(S))
        dset[...] = data
    except Exception:
        dt = h5py.special_dtype(vlen=unicode)
        del handle[name]
        dset = handle.create_dataset(name, (len(data),), dtype=dt)
        dset[...] = data


def load_char_list(char_list_path):
    charlist = []
    with open(char_list_path) as f:
        for l in f:
            charlist.append(l.strip())
    return charlist


def load_file_list_and_transcriptions_and_sizes_and_n_labels(file_list_path, char_list_path, pad_whitespace, base_path):
    charlist = load_char_list(char_list_path)
    file_list = []
    transcription_list = []
    size_list = []
    with open(file_list_path) as f:
        for l in f:
            # IAM format
            if l.startswith("#"):
                continue
            sp = l.split()
            status = sp[1]
            assert status in ("ok", "err"), status
            assert len(sp) >= 9, l
            name = sp[0]
            text = "".join(sp[8:])
            # add space before '
            text = re.sub("([^|])'", "\g<1>|'", text)
            width = int(sp[6])
            height = int(sp[7])
            if height < 1 or width < 1:
                continue
            size_list.append((height, width))
            s = name.split('-')
            name = base_path + name + '.png'
            text = map(lambda c: charlist.index(c), text)
            if pad_whitespace:
                text = [charlist.index("|")] + text + [charlist.index("|")]
            file_list.append(name)
            transcription_list.append(text)
    return file_list, transcription_list, size_list, len(charlist)


def write_to_hdf(file_list, transcription_list, charlist, n_labels, out_file_name, dataset_prefix, pad_y=15, pad_x=15,
                 compress=True):
    with h5py.File(out_file_name, "w") as f:
        f.attrs["inputPattSize"] = 1
        f.attrs["numDims"] = 1
        f.attrs["numSeqs"] = len(file_list)
        classes = charlist

        inputs = []
        sizes = []
        seq_lengths = []
        targets = []
        for i, (img_name, transcription) in enumerate(zip(file_list, transcription_list)):
            targets += transcription
            img = imread(img_name)
            img = 255 - img
            img = numpy.pad(img, ((pad_y, pad_y), (pad_x, pad_x)), 'constant')
            sizes.append(img.shape)
            img = img.reshape(img.size, 1)
            inputs.append(img)
            seq_lengths.append([[img.size, len(transcription), 2]])
            if i % 100 == 0:
                print
                i, "/", len(file_list)

        inputs = numpy.concatenate(inputs, axis=0)
        sizes = numpy.concatenate(numpy.array(sizes, dtype="int32"), axis=0)
        seq_lengths = numpy.concatenate(numpy.array(seq_lengths, dtype="int32"), axis=0)
        targets = numpy.array(targets, dtype="int32")

        f.attrs["numTimesteps"] = inputs.shape[0]

        if compress:
            f.create_dataset("inputs", compression="gzip", data=inputs.astype("float32") / 255.0)
        else:
            f["inputs"] = inputs.astype("float32") / 255.0
        hdf5_strings(f, "labels", classes)
        f["seqLengths"] = seq_lengths
        seq_tags = [dataset_prefix + "/" + tag.split("/")[-1].split(".png")[0] for tag in file_list]
        hdf5_strings(f, "seqTags", seq_tags)

        f["targets/data/classes"] = targets
        f["targets/data/sizes"] = sizes
        hdf5_strings(f, "targets/labels/classes", classes)
        hdf5_strings(f, "targets/labels/sizes", ["foo"])  # TODO, can we just omit it?
        g = f.create_group("targets/size")
        g.attrs["classes"] = len(classes)
        g.attrs["sizes"] = 2


def convert(file_list_path, char_list_path, selections, out_file_names, pad_whitespace, dataset_prefix, base_path,
            compress):
    charlist = load_char_list(char_list_path)
    file_list, transcription_list, size_list, n_labels = load_file_list_and_transcriptions_and_sizes_and_n_labels(
        file_list_path, char_list_path, pad_whitespace, base_path)

    for selection, out_file_name in zip(selections, out_file_names):
        print out_file_name
        selection_set = set(selection)
        assert selection_set.issubset(set(x.split("/")[-1].split(".png")[0] for x in file_list))
        selected_file_list = []
        selected_transcription_list = []
        for f, t in zip(file_list, transcription_list):
            if f.split("/")[-1].split(".png")[0] in selection_set:
                selected_file_list.append(f)
                selected_transcription_list.append(t)
        write_to_hdf(selected_file_list, selected_transcription_list, charlist, n_labels, out_file_name, dataset_prefix,
                     compress)


def convert_IAM_lines_demo(base_path_imgs, tag):
    base_path_out = "features/" + tag + "/"
    mkdir_p(base_path_out)
    file_list_path = "lines.txt"
    char_list_path = "chars.txt"
    out_file_name_demo = base_path_out + "test1.h5"

    print "converting IAM_lines to", out_file_name_demo
    demo_list = ["a01-000u-00"]
    selections = [demo_list]
    out_file_names = [out_file_name_demo]

    convert(file_list_path, char_list_path, selections, out_file_names, pad_whitespace=True, dataset_prefix="trainset",
            base_path=base_path_imgs, compress=False)


def main():
    base_path_imgs = "IAM_lines"
    tag = "raw"

    if base_path_imgs[-1] != "/":
        base_path_imgs += "/"
    convert_IAM_lines_demo(base_path_imgs, tag)


if __name__ == "__main__":
    main()`

2. I created a config file to be used in the decoding using "task": "daemon"
`{
"task": "daemon",
"extract": "log-posteriors",
"output_file": "log-posteriors.h5",
"device": "gpuX",
"multiprocessing": true,
"batching": "default",
"max_seqs": 1,
"batch_size": 600000,
"nadam": 1,
"learning_rate": 0.0005,
"learning_rates": "{1: 0.0005, 25: 0.0003, 35: 0.0001}",
"num_epochs": 80,
"update_batch_size": 2000000,
"reinit": 0,
"log_verbosity": 5,
"train": "features/raw/test1.h5",
"eval": "features/raw/test1.h5",
"num_inputs": 1,
"num_outputs": {"data": [1,2], "classes": [79,1], "sizes": [2,1]},
"network": {
"classes_source": {"class": "source", "data_key": "sizes", "from": [], "n_out": 2, "dtype":"int32"},
"1Dto2D":  {"class": "1Dto2D", "from" : ["data", "classes_source"] },
"conv0":   { "class" : "conv2", "n_features" : 15, "filter" : [3, 3], "pool_size" : [2, 2],  "from" : ["1Dto2D"] },
"mdlstm0": { "class" : "mdlstm", "n_out" : 30, "dropout": 0.25, "from" : ["conv0"] },
"conv1" :  { "class" : "conv2", "n_features" : 45,  "dropout": 0.25, "filter" : [3, 3], "pool_size" : [2, 2],  "from" : ["mdlstm0"] },
"mdlstm1": { "class" : "mdlstm", "n_out" : 60, "dropout": 0.25, "from" : ["conv1"] },
"conv2" :  { "class" : "conv2", "n_features" : 75, "dropout": 0.25, "filter" : [3, 3], "pool_size" : [2, 2],  "from" : ["mdlstm1"] },
"mdlstm2": { "class" : "mdlstm", "n_out" : 90, "dropout": 0.25, "from" : ["conv2"] },
"conv3" :  { "class" : "conv2", "n_features" : 105, "dropout": 0.25, "filter" : [3, 3], "pool_size" : [1, 1],  "from" : ["mdlstm2"] },
"mdlstm3": { "class" : "mdlstm", "n_out" : 120, "dropout": 0.25, "from" : ["conv3"] },
"conv4" :  { "class" : "conv2", "n_features" : 105, "dropout": 0.25, "filter" : [3, 3], "pool_size" : [1, 1],  "from" : ["mdlstm3"] },
"mdlstm4": { "class" : "mdlstm", "n_out" : 120, "dropout": 0.25, "from" : ["conv4"] },
"output":  { "class" : "softmax", "loss": "ctc", "dropout": 0.25, "from" : ["mdlstm4"] }
},

"load": "models/mdlstm_real.080",
"ctc_prior_file": "priors/mdlstm_real",
"model": "models/mdlstm_real",
"log": "log/mdlstm_real_test"
}
`

and I got this on the log httpd listening on port 3333

3. I opened a terminal and tried out this command `curl -X POST http://localhost:3333/classify -H "Content-Type: application/json" -d '{"data":[[-0.7, 0.98],[0.62, 1.3]], "classes" : [0,0]}'`

and what I got is a hash.


I have a few questions on "task": "daemon"

1. Can the classification be done without having to connect to a localhost?
2. What do I do with the hash?
3. What should I expect when I run the task?

Please let me know if I'm not being clear enough with my questions.

**Comment by pvoigtlaender** [comment]
I'm not very familiar with the daemon.
@doetsch can you comment on this?

**Comment by selcouthlyBlue** [comment]
Let me just clarify what I understood about the steps you mention in recognizing the handwriting in a single image.

1. I "convert" the image first into an hdf5 file to serve as the input for the model.
2. I perform forwarding using config_fwd with task forward on this data to produce the posteriors.
3. I decode these posteriors using Kaldi or RASR thus giving me the recognized text.
or
2. I perform forwarding using config_fwd with task daemon on this data to produce the posteriors which are going to be decoded. This will give me the recognized text. 

And here's how I see how it goes:
1. An image containing the words handwritten words "I like trains" is to be converted into an hdf5 to serve as an input for the model.
2. Perform forwarding on this input with task forward to produce the posteriors.
3. I decode these posteriors using Kaldi or RASR thus giving me the recognized text "I like trains".

or
2. I perform forwarding using config_fwd with task daemon on this data to produce the posteriors which are going to be decoded. This will give me the recognized text "I like trains". 

Did I get it right? :O

**Comment by pvoigtlaender** [comment]
Yes, looks right to me.

One thing to keep in mind is that the daemon is an easy solution to quickly get a result, but decoding like this is not as powerful as using RASR or Kaldi,  since the daemon will just do a best path decoding without lexicon and language model.

**Comment by selcouthlyBlue** [comment]
Am I correct in saying that the BestPathDecoder returns the recognized text? If so, is it possible to just use the decoder without having to use daemon? :o

**Comment by interxuxing** [comment]
@pvoigtlaender thank you for your explanation.

"
Afterwards you forward your trained model on this data, which gives you a hdf5 file with posteriors for the image.

Then you run a decoder (e.g. Kaldi, RASR).

It might also be possible in a easier way to run the best path decoder using "task": "daemon".

"


what do you mean by using a decoder Kaldi, RASR, is there any example code to decode the posterior file test.hd5? 
BTW, using "task":"daemon" is a kind of decoder for decoding the posterior file?

Thank you again~
